### PR TITLE
[_] feat: setup workspace

### PR DESCRIPTION
**Pull request by apsantiso**
We need an endpoint to setup the workspace after it has been initialized (it is initialized through the gateway). 

Here we add he required information for the workspace such as name, address, description and also link the owner user to the default team. The workspaceUser for the owner is also configured, for our convenience, the mnemonic is encrypted with the owner's public key, like all users in the workspace.

For this to work with our guard, I also added a check to deny access to users if the workspace is yet to be set up.

### CODE REVIEW

File: src/modules/workspaces/workspaces.usecase.ts [hunk 1]
```diff
@@ -101,8 +101,44 @@ export class WorkspacesUsecases {
...
+      spaceLimit: BigInt(0),
+      driveUsage: BigInt(0),
+      backupsUsage: BigInt(0),
+      key: setupWorkspaceDto.encryptedMnemonic,
+      deactivated: false,
+      createdAt: new Date(),
+      updatedAt: new Date(),
+    });
+
+    await this.workspaceRepository.addUserToWorkspace(workspaceUser);
```
[the remark this answers is not in the document]
Comment (apsantiso) [reply src/modules/workspaces/workspaces.usecase.ts:127]: Sorry this has to be drafted! However you are right, I need to check if workspace.owner === user before this. However, I think i will need to add the owner as workspaceUser in the initialize endpoint so the guard  in controllers can protect this endpoint successfully.

[the remark this answers is not in the document]
Comment (apsantiso) [reply src/modules/workspaces/workspaces.usecase.ts:132]: Solved this by adding an extra check to the guard and fetching the workspace with the respective filters to make sure it is the owner of it. Let me know if you were talking about another thing I missed.

File: src/modules/workspaces/guards/workspaces.guard.ts [hunk 2]
```diff
@@ -73,8 +81,17 @@ export class WorkspaceGuard implements CanActivate {
...
+
+    if (isWorkspaceNotSetup && workspace.isUserOwner(user)) {
+      return true;
+    }
 
-    if (isUserNotInWorkspace || isRequiredOwnerRoleAndUserIsNotOwner) {
+    if (
+      isUserNotInWorkspace ||
+      isRequiredOwnerRoleAndUserIsNotOwner ||
+      isWorkspaceNotSetup
```
Comment (sg-gs) [inline src/modules/workspaces/guards/workspaces.guard.ts:93]: This is a bit misleading. In reality, you are rejecting when the workspace is not set up and the user is not the owner.

File: src/modules/workspaces/workspaces.usecase.spec.ts [hunk 3]
```diff
@@ -293,6 +293,83 @@ describe('WorkspacesUsecases', () => {
...
+      jest
+        .spyOn(workspaceRepository, 'findOne')
+        .mockResolvedValueOnce(workspace);
+
+      await service.setupWorkspace(user, 'workspace-id', {
+        name: 'Test Workspace',
+        encryptedMnemonic: 'encryptedMnemonic',
+      });
+
+      expect(workspaceRepository.addUserToWorkspace).toHaveBeenCalled();
```
Comment (sg-gs) [inline src/modules/workspaces/workspaces.usecase.spec.ts:322]: Ensuring the function has been called with the owner's data would make this test stronger

File: src/modules/workspaces/workspaces.usecase.spec.ts [hunk 4]
```diff
@@ -293,6 +293,83 @@ describe('WorkspacesUsecases', () => {
...
+      jest
+        .spyOn(workspaceRepository, 'findOne')
+        .mockResolvedValueOnce(workspace);
+
+      await service.setupWorkspace(user, 'workspace-id', {
+        name: 'Test Workspace',
+        encryptedMnemonic: 'encryptedMnemonic',
+      });
+
+      expect(teamRepository.addUserToTeam).toHaveBeenCalledWith(
```
Comment (sg-gs) [inline src/modules/workspaces/workspaces.usecase.spec.ts:337]: Same here

File: src/modules/workspaces/workspaces.usecase.ts [hunk 5]
```diff
@@ -101,8 +101,49 @@ export class WorkspacesUsecases {
...
+    setupWorkspaceDto: SetupWorkspaceDto,
+  ) {
+    const workspace = await this.workspaceRepository.findOne({
+      id: workspaceId,
+      ownerId: user.uuid,
+      setupCompleted: false,
+    });
+
+    if (!workspace) {
+      throw new NotFoundException('There is no workspace to be setup');
```
Comment (sg-gs) [inline src/modules/workspaces/workspaces.usecase.ts:116]: Workspace not found would be better

File: src/modules/workspaces/workspaces.usecase.ts [hunk 6]
```diff
@@ -101,8 +101,49 @@ export class WorkspacesUsecases {
...
+      spaceLimit: BigInt(0),
+      driveUsage: BigInt(0),
+      backupsUsage: BigInt(0),
+      key: setupWorkspaceDto.encryptedMnemonic,
+      deactivated: false,
+      createdAt: new Date(),
+      updatedAt: new Date(),
+    });
+
+    await this.workspaceRepository.addUserToWorkspace(workspaceUser);
```
Comment (sg-gs) [inline src/modules/workspaces/workspaces.usecase.ts:132]: If this step fails, we may want to check which step is still failing, so in the case of a massive failure, we can just fix the error and users will only need to retry again. What do you think?

**Review by sg-gs — CHANGES_REQUESTED** [review]

**Review by sg-gs — APPROVED** [review]

**Pull request merged**